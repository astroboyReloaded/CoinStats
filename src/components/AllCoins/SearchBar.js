import { useSelector, useDispatch } from 'react-redux';
import { setSearch } from '../../redux/all-coins/allCoinsSlice';
import '../../styles/SearchBar.css';

const SearchBar = () => {
  const handleSubmit = (e) => e.preventDefault();

  const { allCoins: data, search } = useSelector((state) => state.allCoins);
  const dispatch = useDispatch();

  const handleChange = (e) => {
    const search = e.target.value;
    const results = data !== 'All Coins' ? data.filter((item) => (
      item.id.indexOf(search.toLowerCase()) > -1
      || item.symbol.indexOf(search.toLowerCase()) > -1)) : [];
    const searchResults = {
      search,
      results,
    };
    dispatch(
      setSearch(searchResults),
    );
  };

  return (
    <form onSubmit={handleSubmit} className="searchContainer">
      <input
        onChange={(e) => handleChange(e)}
        type="text"
        placeholder="Search"
        value={search?.length ? search : ''}
        className="searchBar"
      />
    </form>
  );
};

export default SearchBar;
