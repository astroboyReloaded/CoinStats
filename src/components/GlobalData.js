import { useSelector, useDispatch } from 'react-redux';
import { useEffect } from 'react';
import {
  fetchGlobalData,
  clearGlobalData,
} from '../redux/global-data/globalDataSlice';

const GlobalData = () => {
  const { globalData, isLoading, error } = useSelector(
    (state) => state.globalData,
  );
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchGlobalData());
    return () => dispatch(clearGlobalData());
  }, [dispatch]);

  if (isLoading) {
    return <p>Loading...</p>;
  }
  if (error) {
    return <p>Ups! Something went wrong...</p>;
  }
  if (globalData !== 'Global Data') {
    const { data } = globalData;
    return (
      <main>
        <p>{`GLOBAL MARKET CAP: $${data.total_market_cap.usd}`}</p>
        <p>{`24HR VOLUME: $${data.total_volume.usd}`}</p>
      </main>
    );
  }
  return <h1>GlobalData</h1>;
};

export default GlobalData;
