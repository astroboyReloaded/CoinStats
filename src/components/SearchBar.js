import { useSelector, useDispatch } from 'react-redux';
import { setSearch } from '../redux/all-coins/allCoinsSlice';

const SearchBar = () => {
  const handleSubmit = (e) => e.preventDefault();

  const { allCoins: data, search } = useSelector((state) => state.allCoins);
  const dispatch = useDispatch();

  const handleChange = (e) => {
    const search = e.target.value.toLowerCase();
    const results = data.filter((item) => (
      item.id.indexOf(search) > -1
      || item.symbol.indexOf(search) > -1));
    const searchResults = {
      search,
      results,
    };
    dispatch(
      setSearch(searchResults),
    );
  };

  const handleClear = () => {
    dispatch(setSearch({
      search: '',
      results: [],
    }));
  };

  return (
    <form onSubmit={handleSubmit}>
      <input
        onChange={(e) => handleChange(e)}
        type="text"
        placeholder="Search"
        value={search?.length ? search : ''}
      />
      <input type="button" value="â†»" onClick={handleClear} />
    </form>
  );
};

export default SearchBar;
