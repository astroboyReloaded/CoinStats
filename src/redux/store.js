import { configureStore } from '@reduxjs/toolkit';
import allCoinsReducer from './all-coins/allCoinsSlice';
import coinDetailsReducer from './coin-details/coinDetailsSlice';
import categoriesReducer from './categories/categoriesSlice';
import globalDataReducer from './global-data/globalDataSlice';

const store = configureStore({
  reducer: {
    allCoins: allCoinsReducer,
    coinDetails: coinDetailsReducer,
    categories: categoriesReducer,
    globalData: globalDataReducer,
  },
});

export default store;
