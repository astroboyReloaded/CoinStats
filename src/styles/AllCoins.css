.All-Loading {
  color: gainsboro;
}

.categoriesMain {
  font-family: Arial, Helvetica, sans-serif;
}

.coin-table {
  margin-top: 0.3rem;
  display: flex;
  flex-direction: column;
  color: goldenrod;
}

.head-tr {
  display: flex;
  justify-content: space-between;
  margin-inline: 0.3rem;
  font-size: 0.72rem;
  border-bottom: 1px solid goldenrod;
  border-top: 1px solid goldenrod;
}

h3 { margin: 0; }

.coin-tbody {
  display: flex;
  flex-direction: column;
  margin-inline: 0.3rem;
  font-size: 0.81rem;
}

.body-tr {
  display: flex;
  justify-content: space-between;
  padding: 0.3rem 0;
  align-items: center;
  color: gainsboro;
  border-bottom: 1px solid goldenrod;
}

p { margin: 0; }

.rank {
  min-width: 1.2rem;
  text-align: left;
}

.coin-td {
  min-width: 3rem;
  text-align: center;
}

.coin-img {
  max-width: 0.9rem;
}

.all-coins-symbol {
  font-size: 0.6rem;
  margin: 0;
}

.price-td {
  min-width: 3.9rem;
  text-align: right;
}

.perc-24h-td {
  min-width: 3.3rem;
  text-align: right;
}

.mrkt-cap-td {
  min-width: 7.2rem;
  text-align: right;
}
