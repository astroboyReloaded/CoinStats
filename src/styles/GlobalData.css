.global-data {
  display: flex;
  margin: 0;
  padding: 0.3rem;
  color: gainsboro;
  font-family: Arial, Helvetica, sans-serif;
}

.global-data-title {
  border-radius: 6px;
  border-left: 6px solid yellowgreen;
  font-size: 0.72rem;
  text-align: left;
  padding-inline: 0.3rem;
  background-color: #20223f;
}

.marketCap {
  margin-right: 0.3rem;
}
