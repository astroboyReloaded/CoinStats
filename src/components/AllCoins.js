import { useSelector, useDispatch } from 'react-redux';
import { useEffect } from 'react';
import { Link } from 'react-router-dom';
import SearchBar from './SearchBar';
import { fetchAllCoins } from '../redux/all-coins/allCoinsSlice';
import { setupTop100 } from '../redux/price-convertion/priceConvertionSlice';
import GlobalData from './GlobalData';
import '../styles/AllCoins.css';

const AllCoins = () => {
  const {
    allCoins,
    results,
    isLoading,
    error,
  } = useSelector(
    (state) => state.allCoins,
  );
  const dispatch = useDispatch();

  useEffect(() => {
    if (allCoins !== 'All Coins') return;
    dispatch(fetchAllCoins());
  }, [dispatch, allCoins]);

  const handleSetTop100CoinsPrices = () => {
    const convertionCurrencies = allCoins.map((coin) => ({
      image: coin.image,
      symbol: coin.symbol,
      price: coin.current_price,
    }));
    dispatch(setupTop100(convertionCurrencies));
  };

  if (isLoading) {
    return <p className="All-Loading">Loading...</p>;
  }

  if (error) {
    return <p>{error}</p>;
  }

  if (allCoins !== 'All Coins') {
    const render = results === null ? allCoins : results;
    console.log(JSON.stringify(render));
    return (
      <main className="categoriesMain">
        <GlobalData />
        <SearchBar />
        <section className="coin-table">
          <header className="head-tr">
            <h3 className="rank">#</h3>
            <h3 className="coin-td">COIN</h3>
            <h3 className="price-td">PRICE</h3>
            <h3 className="perc-24h-td">24H</h3>
            <h3 className="mrkt-cap-td">MARKET CAP</h3>
          </header>
          <main className="coin-tbody">
            {render.map((coin) => (
              <Link to={`coin-details/${coin.id}`} onClick={handleSetTop100CoinsPrices} key={coin.id}>
                <article
                  className="body-tr"
                  data-testid="coinData-Container"
                >
                  <p className="rank">{coin.market_cap_rank}</p>
                  <p className="coin-td">
                    <img
                      src={coin.image}
                      alt={coin.name}
                      className="coin-img"
                    />
                    <h2 className="all-coins-symbol">{coin.symbol.toUpperCase()}</h2>
                  </p>
                  <p className="price-td">
                    <data className="coin-price">{`$${coin.current_price.toLocaleString()}`}</data>
                  </p>
                  <p className="perc-24h-td">
                    <data className={`prc-chng-perc ${coin.price_change_percentage_24h > 0 ? 'green' : 'red'}`}>{`${coin.price_change_percentage_24h.toFixed(1).replace('-', '')}%`}</data>
                  </p>
                  <p className="mrkt-cap-td">
                    <data>{`$${coin.market_cap?.toLocaleString()}`}</data>
                  </p>
                </article>
              </Link>
            ))}
          </main>
        </section>
      </main>
    );
  }

  return allCoins;
};

export default AllCoins;
