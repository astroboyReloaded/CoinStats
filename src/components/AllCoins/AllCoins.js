import SearchBar from '../SearchBar';
import GlobalData from '../GlobalData';
import CoinArticleLink from './CoinArticleLink';
import '../../styles/AllCoins.css';
import useFetchAllCoins from '../../hooks/allCoins-hooks/useFetchAllCoins';

const AllCoins = () => {
  const [allCoins, results, isLoading, error] = useFetchAllCoins();

  if (isLoading) {
    return <p className="All-Loading">Loading...</p>;
  }

  if (error) {
    return <p>{error}</p>;
  }

  if (allCoins !== 'All Coins') {
    const render = results === null ? allCoins : results;
    return (
      <main className="categoriesMain">
        <GlobalData />
        <SearchBar />
        <section className="coin-table">
          <header className="head-tr">
            <h3 className="rank">#</h3>
            <h3 className="coin-td">COIN</h3>
            <h3 className="price-td">PRICE</h3>
            <h3 className="perc-24h-td">24H</h3>
            <h3 className="mrkt-cap-td">MARKET CAP</h3>
          </header>
          <main className="coin-tbody">
            {render.map((coin) => (
            <CoinArticleLink
                key={coin.id}
                id={coin.id}
                mrktCapRank={coin.market_cap_rank}
                image={coin.image}
                name={coin.name}
                symbol={coin.symbol}
                currentPrice={coin.current_price}
                priceChngPerc24h={coin.price_change_percentage_24h}
                marketCap={coin.market_cap}
              />
            ))}
          </main>
        </section>
      </main>
    );
  }

  return allCoins;
};

export default AllCoins;
