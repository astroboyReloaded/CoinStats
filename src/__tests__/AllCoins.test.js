import renderer from 'react-test-renderer';
import { BrowserRouter, Routes, Route } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../redux/store';
import AllCoins from '../components/AllCoins/AllCoins';
import GlobalData from '../components/AllCoins/GlobalData';
import SearchBar from '../components/AllCoins/SearchBar';

it('renders AllCoins component', () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <Provider store={store}>
          <Routes>
            <Route path="/">
              <Route index element={<AllCoins />} />
            </Route>
          </Routes>
        </Provider>
      </BrowserRouter>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders <SearchBar> and <GlobalData> components', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <AllCoins>
        <GlobalData />
        <SearchBar />
      </AllCoins>
    </Provider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
