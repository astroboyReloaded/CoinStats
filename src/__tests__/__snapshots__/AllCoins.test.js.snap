// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders <SearchBar> and <GlobalData> components 1`] = `
<div
  className="boxes"
>
  <div
    className="box"
  >
    <div />
    <div />
    <div />
    <div />
  </div>
  <div
    className="box"
  >
    <div />
    <div />
    <div />
    <div />
  </div>
  <div
    className="box"
  >
    <div />
    <div />
    <div />
    <div />
  </div>
  <div
    className="box"
  >
    <div />
    <div />
    <div />
    <div />
  </div>
</div>
`;

exports[`renders AllCoins component 1`] = `"All Coins"`;
